### PR TITLE
add tag-retention tab to project detail

### PR DESCRIPTION
**Pull request by AllForNothing**
#8360 
Signed-off-by: Shijun Sun (c) <sshijun@vmware.com>
image
image

### CODE REVIEW

File: src/portal/src/app/project/project.module.ts [hunk 1]
```diff
@@ -39,6 +39,10 @@ import { HelmChartModule } from './helm-chart/helm-chart.module';
 import { RobotAccountComponent } from './robot-account/robot-account.component';
 import { AddRobotComponent } from './robot-account/add-robot/add-robot.component';
 import { AddHttpAuthGroupComponent } from './member/add-http-auth-group/add-http-auth-group.component';
+import {TagRetentionComponent} from "./tag-retention/tag-retention.component";
+import {AddRuleComponent} from "./tag-retention/add-rule/add-rule.component";
+import {TagRetentionService} from "./tag-retention/tag-retention.service";
+import {EditRuleComponent} from "./tag-retention/edit-rule/edit-rule.component";
```
Comment (steven-zou) [inline src/portal/src/app/project/project.module.ts]: WHY use different component `EditRuleComponent` and `AddRuleComponent` to cover the rule adding and editing?

File: src/portal/src/app/project/tag-retention/add-rule/add-rule.component.html [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            <div class="clr-row">
+                <div class="clr-col-4">
+                    <div class="over-line"></div>
+                    <label>{{'TAG_RETENTION.BY_WHAT' | translate}}</label>
+                </div>
+                <div class="clr-col-5">
+                    <div class="over-line"></div>
+                    <div class="clr-select-wrapper w-100">
+                        <select [(ngModel)]="template" class="clr-select w-100">
+                            <!--<option *ngFor="let t of metadata?.templates" value="{{t?.rule_template}}">{{t?.display_text}}</option>-->
```
Comment (steven-zou) [inline src/portal/src/app/project/tag-retention/add-rule/add-rule.component.html]: WHY comment the loop?

File: src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    selector: "add-rule",
+    templateUrl: "./add-rule.component.html",
+    styleUrls: ["./add-rule.component.scss"]
+})
+export class AddRuleComponent implements OnInit, OnDestroy {
+
+    addRuleOpened: boolean = false;
+    @Output() clickAdd = new EventEmitter<Rule>();
+    template: string;
+    number: number;
```
Comment (steven-zou) [inline src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts]: Maybe you can define an interface to cover all those data

File: src/portal/src/app/project/tag-retention/edit-rule/edit-rule.component.html [hunk 4]
```diff
@@ -0,0 +1,110 @@
+<clr-modal [(clrModalOpen)]="editRuleOpened"
```
Comment (steven-zou) [inline src/portal/src/app/project/tag-retention/edit-rule/edit-rule.component.html]: You can use the add dialog component to do this

File: src/portal/src/app/project/tag-retention/tag-retention.component.html [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        <label class="fw9-l">{{'TAG_RETENTION.RETENTION_RULES' | translate}}</label><span class="badge badge-3 clr-offset-1">{{retention?.rules?.length ? retention?.rules?.length : 0}}</span>
+    </div>
+    <div class="clr-col">
+        <ul *ngIf="retention?.rules?.length > 0" class="list-unstyled">
+            <li class="rule" *ngFor="let rule of retention?.rules;let i = index;">
+                <div class="clr-row">
+                    <div class="clr-col-11 rule-name">
+                        <!--<clr-icon *ngIf="!rule?.isDisabled" shape="success-standard" class="color-green"></clr-icon>
+                        <clr-icon *ngIf="rule?.isDisabled" shape="error-standard" class="color-red"></clr-icon>-->
+                        <span *ngIf="rule?.template==='lastXDays'">{{'TAG_RETENTION.RULE_NAME_1'|translate:{number:rule?.params?.num,rep:repositoryToString(rule?.scope_selectors?.repository)} }}</span>
```
Comment (steven-zou) [inline src/portal/src/app/project/tag-retention/tag-retention.component.html]: This seems wrong. You should use an array loop to render the list

File: src/portal/src/app/project/tag-retention/tag-retention.service.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        .pipe(map(response => response as Array<any>))
+        .pipe(catchError(error => observableThrowError(error)));
+  }
+  getExecutionHistory(retentionId, executionId) {
+    return this.http.get(`/api/retentions/${retentionId}/executions/${executionId}/tasks`)
+        .pipe(map(response => response as Array<any>))
+        .pipe(catchError(error => observableThrowError(error)));
+  }
+  seeLog(retentionId, executionId, taskId) {
+    window.open(`api/retention/${retentionId}/executions/${executionId}/tasks/${taskId}`, '_blank');
```
Comment (steven-zou) [inline src/portal/src/app/project/tag-retention/tag-retention.service.ts]: WHAT's this?

File: src/portal/src/app/user/user.service.ts [hunk 7]
```diff
@@ -47,9 +47,11 @@ export class UserService {
             , catchError(error => this.handleError(error)));
     }
     getUsers(): Observable<User[]> {
-        return this.http.get(userMgmtEndpoint, HTTP_GET_OPTIONS)
-            .pipe(map(response => response as User[])
-            , catchError(error => this.handleError(error)));
+        return this.http.get(userMgmtEndpoint)
+            .pipe(map(((response: any) => {
+                console.log(response.headers.get('X-Total-Count'));
```
Comment (jwangyangls) [inline src/portal/src/app/user/user.service.ts]: delete it

File: src/portal/src/app/project/tag-retention/add-rule/add-rule.component.html [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            <div class="clr-row">
+                <div class="clr-col-4">
+                    <div class="over-line"></div>
+                    <label>{{'TAG_RETENTION.BY_WHAT' | translate}}</label>
+                </div>
+                <div class="clr-col-5">
+                    <div class="over-line"></div>
+                    <div class="clr-select-wrapper w-100">
+                        <select [(ngModel)]="template" class="clr-select w-100">
+                            <option *ngFor="let t of metadata?.templates" value="{{t.rule_template}}">{{getI18nKey(t?.display_text)|translate}}</option>
```
[the remark this answers is not in the document]
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/add-rule/add-rule.component.html]: fixed

File: src/portal/src/app/project/tag-retention/tag-retention.component.html [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+                <clr-dg-string-filter [clrDgStringFilter]="dryRunFilter"></clr-dg-string-filter>
+            </clr-dg-column>
+            <clr-dg-column>
+                {{'TAG_RETENTION.START_TIME' | translate}}
+            </clr-dg-column>
+            <clr-dg-column>
+                {{'TAG_RETENTION.DURATION' | translate}}
+            </clr-dg-column>
+            <!-- <clr-dg-column>
+                 {{'TAG_RETENTION.DETAILS' | translate}}
```
[the remark this answers is not in the document]
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/tag-retention.component.html]: fixed

File: src/portal/src/app/project/tag-retention/tag-retention.service.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    return this.http.post(`/api/retentions/${retentionId}/executions`, {dry_run: true})
+        .pipe(catchError(error => observableThrowError(error)));
+  }
+  AbortRun(retentionId, executionId) {
+    return this.http.patch(`/api/retentions/${retentionId}/executions/${executionId}`, {action: 'stop'})
+        .pipe(catchError(error => observableThrowError(error)));
+  }
+  getRunNowList(retentionId) {
+    return this.http.get(`/api/retentions/${retentionId}/executions`)
+        .pipe(map(response => response as Array<any>))
```
[the remark this answers is not in the document]
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/tag-retention.service.ts]: no specific type

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+
+    editRuleByIndex(index) {
+        this.editIndex = index;
+        this.addRuleComponent.rule = clone(this.retention.rules[index]);
+        this.addRuleComponent.editRuleOrigin = clone(this.retention.rules[index]);
+        this.addRuleComponent.open();
+        this.addRuleComponent.isAdd = false;
+        this.ruleIndex = -1;
+    }
+    disableRule(index) {
```
[the remark this answers is not in the document]
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/tag-retention.component.ts]: deleted

File: src/portal/src/app/user/user.service.ts (hunk 7 above)
Comment (AllForNothing) [reply src/portal/src/app/user/user.service.ts]: fixed

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+                this.retention = retention;
+            }, error => {
+                this.errorHandler.error(error);
+            });
+    }
+
+    deleteRule(index) {
+        let retention: Retention = clone(this.retention);
+        retention.rules.splice(index, 1);
+        this.ruleIndex = -1;
```
[the remark this answers is not in the document]
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/tag-retention.component.ts]: delete rule from a cloned ruleList , if api return success, then set value to this.ruleList

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts [hunk 13]
```diff
@@ -0,0 +1,306 @@
+// Copyright Project Harbor Authors
```
[the remark this answers is not in the document]
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/tag-retention.component.ts]: fixed

File: src/portal/src/app/project/tag-retention/tag-retention.component.html [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+                        </span>
+                    </div>
+                    <div class="clr-col-1">
+                        <clr-icon (click)="openEditor(i)" shape="ellipsis-vertical"
+                                  class="dropdown-toggle hand"></clr-icon>
+                        <div class="dropdown" [ngClass]="{open:ruleIndex===i}">
+                            <div class="dropdown-menu">
+                                <button type="button" class="dropdown-item"
+                                        (click)="editRuleByIndex(i)">{{'TAG_RETENTION.EDIT' | translate}}</button>
+                                <!--<button *ngIf="!rule?.isDisabled" type="button" class="dropdown-item" (click)="disableRule(i)">{{'TAG_RETENTION.DISABLE' | translate}}</button>
```
[the remark this answers is not in the document]
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/tag-retention.component.html]: fixed

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+        accepts(item: any, search: string): boolean {
+            let str = item.dry_run ? 'YES' : 'NO';
+            return str.indexOf(search) !== -1;
+        }
+    };
+    projectId: number;
+    addRuleOpened: boolean = false;
+    isRetentionRunOpened: boolean = false;
+    isAbortedOpened: boolean = false;
+    selectedItem;
```
[the remark this answers is not in the document]
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/tag-retention.component.ts]: fixed

File: src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts [hunk 16]
```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2017 VMware, Inc. All Rights Reserved.
```
Comment (zhoumeina) [inline src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts]: All the ts file need to add license now? Why the license is 2017?

File: src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+
+    ngOnDestroy(): void {
+    }
+    set template(template) {
+        this.rule.template = template;
+    }
+    get template() {
+        return this.rule.template;
+    }
+     get unit(): string {
```
Comment (zhoumeina) [inline src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts]: format this unit() method

File: src/portal/src/app/project/tag-retention/retention.ts [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+export class Rule {
+    isDisabled: boolean;
+    id: number;
+    priority: number;
+    action: string;
+    template: string;
+    params: {
+        num: number
+    };
+    tag_selectors: Array<{kind: string; decoration: string; pattern: string}>;
```
Comment (zhoumeina) [inline src/portal/src/app/project/tag-retention/retention.ts]: Could you please define a class for: {kind: string; decoration: string; pattern: string}. Then you can use that class here.

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts [hunk 19]
```diff
@@ -0,0 +1,282 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (zhoumeina) [inline src/portal/src/app/project/tag-retention/tag-retention.component.ts]: This license is different from the previous one.

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts [hunk 20]
```diff
@@ -0,0 +1,282 @@
...
+            this.tagRetentionService.getRetention(this.retentionId).subscribe(
+                response => {
+                    this.retention = response;
+                }, error => {
+                    this.errorHandler.error(error);
+                });
+        }
+    }
+
+    repositoryToString(repositories): string {
```
Comment (jwangyangls) [inline src/portal/src/app/project/tag-retention/tag-retention.component.ts]: Where is it used?

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (jwangyangls) [reply src/portal/src/app/project/tag-retention/tag-retention.component.ts]: if the api failed ,what  should this.ruleIndex be in?

File: src/portal/src/app/project/tag-retention/retention.ts (hunk 18 above)
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/retention.ts]: modified

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+        this.addRuleComponent.editRuleOrigin = clone(this.retention.rules[index]);
+        this.addRuleComponent.open();
+        this.addRuleComponent.isAdd = false;
+        this.ruleIndex = -1;
+    }
+    deleteRule(index) {
+        let retention: Retention = clone(this.retention);
+        retention.rules.splice(index, 1);
+        this.ruleIndex = -1;
+        this.tagRetentionService.updateRetention(this.retentionId, retention).subscribe(
```
Comment (jwangyangls) [inline src/portal/src/app/project/tag-retention/tag-retention.component.ts]: why is no loading when update list?

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+            retention.rules.push(rule);
+            if (!this.retentionId) {
+                this.tagRetentionService.createRetention(retention).subscribe(
+                    response => {
+                        this.refreshAfterCreatRetention();
+                    }, error => {
+                        this.errorHandler.error(error);
+                    });
+            } else {
+                this.tagRetentionService.updateRetention(this.retentionId, retention).subscribe(
```
Comment (jwangyangls) [inline src/portal/src/app/project/tag-retention/tag-retention.component.ts]: why is no loading when update list?

Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/tag-retention.component.ts]: not neccesary

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts (hunk 21 above)
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/tag-retention.component.ts]: not necesary

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts (hunk 20 above)
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/tag-retention.component.ts]: deleted

File: src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts (hunk 16 above)
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts]: modified

File: src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+// limitations under the License.
+import {
+    Component,
+    OnInit,
+    Input,
+    OnDestroy,
+    Output,
+    EventEmitter,
+} from "@angular/core";
+import {Rule, RuleMetadate} from "../retention";
```
Comment (zhoumeina) [inline src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts]: Please format this file.

File: src/portal/src/app/project/tag-retention/retention.ts [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+    decoration: string;
+    pattern: string;
+}
+
+export class RuleMetadate {
+    templates: Array<{
+        rule_template: string;
+        display_text: string;
+        action: "retain",
+        params: Array<{
```
Comment (zhoumeina) [inline src/portal/src/app/project/tag-retention/retention.ts]: The same here.

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts (hunk 21 above)
Comment (jwangyangls) [reply src/portal/src/app/project/tag-retention/tag-retention.component.ts]: I think it's necessary.

File: src/portal/src/app/project/tag-retention/retention.ts (hunk 24 above)
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/retention.ts]: modified

File: src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts (hunk 23 above)
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/add-rule/add-rule.component.ts]: fixed

File: src/portal/src/app/project/tag-retention/tag-retention.component.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (AllForNothing) [reply src/portal/src/app/project/tag-retention/tag-retention.component.ts]: "-1" is origin value

**Pull request merged**